### PR TITLE
[Bugfix] Fix Jenkins fix localhost in pipeline

### PR DESCRIPTION
**Pull request by ivanchimeno**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context

It's currently not possible to get Artemis up and running with the Gitlab/Jenkins environment on Mac OS X (and possibly Windows). On Docker for Mac, the containers are only accessible via `localhost`. 

This leads to two issues when running the environment locally:

- When using Jenkins as a local agent, it cannot clone repositories from Gitlab because `localhost` is not the address of the Gitlab container
- The Gitlab to Jenkins webhook on push events doesn't work because `localhost` is not the address of the Jenkins container.

### Description

The pr adds two new entries `jenkins.docker.vcs-url` and `jenkins.docker.ci-url` in `application-jerkins.yml` enabling the user two enter the ip addresses of the docker containers. The configured urls are then used for the Jenkins Pipeline scripts and webhook.

### Steps for Testing

1. Mac users should be able to successfully get Artemis up and running with Jenkins and Gitlab

### Test Coverage

JenkinsDockerUrlService.java 100%
JenkinsService.java 82%

### CODE REVIEW

**Comment by dfuchss** [comment]
Just a remark / question:

Would it help, if you set alias values to a common nginx proxy that delegates to the actual containers in a docker-compose:
e.g.:

```
nginx:
    image: nginx:latest
    ports:
      - "80:80"
      - "443:443"
    networks:
      artemis-net:
        aliases:
          - ${ARTEMIS_SERVER_NAME}
          - ${GIT_SERVER_NAME}
          - ${JENKINS_SERVER_NAME}

```

You wouldn't have to use `localhost` as name of the services and therefore have no problems regarding cors in local development. You should be able to use jenkins , gitlab , etc. using the actual ports but internally all traffic could be routed within a virtual network.

I don't know if this is off topic, but maybe this would work on windows?

File: src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsDockerUrlService.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import org.springframework.web.util.UriComponentsBuilder;
+
+import de.tum.in.www1.artemis.domain.VcsRepositoryUrl;
+
+@Profile("jenkins")
+@Service
+public class JenkinsDockerUrlService {
+
+    @Value("${jenkins.docker.ci-url:#{null}}")
+    private URL dockerCiUrl;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsDockerUrlService.java:20]: I would propose to use Optional here

File: src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsDockerUrlService.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    @Value("${jenkins.docker.vcs-url:#{null}}")
+    private URL dockerVcsUrl;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsDockerUrlService.java:23]: I would propose to use Optional here

**Comment by krusche** [comment]
Please make sure the test coverage for the changed files is > 80%

**Review by SimonEntholzer — DISMISSED** [review]
The code and the test look good to me, I couldn't test it though because I do not have Jenkins running locally. ( and also only have linux, so it would have worked for me even without the fix )

File: src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsDockerUrlService.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    /**
+     * Returns the url pointing the the docker container of the ci
+     * if the ci url is set. Returns the original url otherwise.
+     * @param ciUrl The url of the ci
+     * @return the manipulated url
+     */
+    public String toDockerCiUrl(String ciUrl) {
+        if (dockerCiUrl.isPresent()) {
```
Comment (jamalxo) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsDockerUrlService.java:61]: maybe use ternary condition instead and write it in one line

File: src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsDockerUrlService.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        if (dockerVcsUrl.isEmpty()) {
+            return vcsRepositoryUrl;
+        }
+
+        if (vcsRepositoryUrl.getURL() == null) {
+            log.warn("Cannot replace url to docker url {} because the url is null.", dockerVcsUrl);
+            return vcsRepositoryUrl;
+        }
+
+        try {
```
Comment (jamalxo) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsDockerUrlService.java:44]: cant we catch this error case beforehand, like you did with the others above?

Comment (ivanchimeno) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsDockerUrlService.java:44]: The exception happens on line 46 so only when I create and return the object.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsDockerUrlService.java (hunk 3 above)
Comment (ivanchimeno) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsDockerUrlService.java:61]: I think it's a matter of preference :D

Comment (jamalxo) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsDockerUrlService.java:61]: it looks cleaner that way but we can keep it like it is

**Review by jamalxo — APPROVED** [review]
code looks fine and tested locally

**Comment by krusche** [comment]
It would be nice if 2-3 more people with a local Jenkins setup could test if this is working properly without any issues

**Comment by LDAP** [comment]
I don't quite understand why these changes are necessary. I use a local setup with Jenkins on Windows and have set as CI url `jenkins:port`, where `jenkins` is the name of the service. The internal docker DNS then resolves this name in the internal docker network. Additionally, I do have published the Jenkins web-server port onto localhost to access it with my browser.

**Comment by dfuchss** [comment]
I strongly agree on the point of @LDAP. Actually, that's the procedure I'm using on Windows as well:

Artemis communicates with its services within the artemis virtual network. So you can use `container-name:-real-port-at-container` as endpoint.
In addition, as Lucas already mentioned, you can remap these ports to access the services from your browser (but this mapping is independent from the actual artemis communication within the virtual network).

**Comment by dfuchss** [comment]
> You wouldn't have to use `localhost` as name of the services and therefore have no problems regarding cors in local development. You should be able to use jenkins , gitlab , etc. using the actual ports but internally all traffic could be routed within a virtual network.

Actually, that's what I have mentioned here :D

**Comment by dfuchss** [comment]
That are some excerpts from our config for https://github.com/kit-sdq/Artemis-SAML2-Test-Docker
We use defined ports to access jenkins and gitlab from our browser; but artemis uses the internal network.

application-artemis.yml:
```
artemis:
    version-control:
        url: http://gitlab:80
    continuous-integration:
        url: http://jenkins:8080        
 ```

docker-compose.yml:

```
version: '3'
services:
  gitlab:
    ports:
      - "${GITLAB_HTTP_PORT}:80"
      - "${GITLAB_HTTPS_PORT}:443"
      - "${GITLAB_SSH_PORT}:22"
    networks:
      - artemis-net

  jenkins:
    ports:
      - "50000:50000"
      - "${JENKINS_HTTP_PORT}:8080"
    networks:
      - artemis-net

networks:
  artemis-net:

```

**Comment by b-fein** [comment]
If I understand the issue correctly, it is one I stumbled upon before as well:
Before Artemis used the URLs provided in `artemis.{version_control,continuous-integration}.url` both to generate the links in the client to the repository (e.g. clone link) and the webhooks between Gitlab and Jenkins. This led to issues for me:
- If I set `artemis.version_control.url: http://localhost:8082`: Link in Client works, but Jenkins building is broken.
- If I set `artemis.version_control.url: http://gitlab:80`: Jenkins building works, but the link in Client (e.g. clone URL) is broken as the hostname from inside the docker network is not known in the browser (unless I add it to `/etc/hosts`).
- The only easy solution for me was to use the actual (forced to be static) IPs of the containers e.g. `artemis.version_control.url: http://172.33.0.2` that are then reachable both from client and Jenkins container.

Now those two uses of the URLs are separated and Jenkins/Gitlab can communicate completely inside the docker network with each other via their hostnames and possibly changing IPs while the client uses their externally available URLs.

Therefore this fixes not only the problem for the local setup in docker, but also production use cases where the communication between Gitlab and Jenkins is different to the one of each of the services to the outside.

**Review by b-fein — DISMISSED** [review]
I tested it locally with the recently introduced new docker-compose setup by now uncommenting the two options in `application-jenkins.yml` and setting the URLs of Gitlab and Jenkins to their respective `http://localhost:$port` in `application-local.yml`.

After creating a new exercise the build in Jenkins was triggered and cloning the repositories worked. Therefore, the communication between the services worked as intended.

**Comment by dfuchss** [comment]
Ok. Now I got the actual Issue :) Thanks for clarification :)

**Comment by LDAP** [comment]
> **RFC**: Should `jenkins.docker` maybe be renamed to something like `jenkins.internal-urls` in `application-jenkins.yml` with an updated comment as well to reflect that those settings may not only be useful in a local setup? Then the `JenkinsDockerUrlService` and related functions should get a new name, too.

I strongly agree with this suggesting.

**Comment by ivanchimeno** [comment]
> If I understand the issue correctly, it is one I stumbled upon before as well:
> Before Artemis used the URLs provided in `artemis.{version_control,continuous-integration}.url` both to generate the links in the client to the repository (e.g. clone link) and the webhooks between Gitlab and Jenkins. This led to issues for me:
> 
> * If I set `artemis.version_control.url: http://localhost:8082`: Link in Client works, but Jenkins building is broken.
> * If I set `artemis.version_control.url: http://gitlab:80`: Jenkins building works, but the link in Client (e.g. clone URL) is broken as the hostname from inside the docker network is not known in the browser (unless I add it to `/etc/hosts`).
> * The only easy solution for me was to use the actual (forced to be static) IPs of the containers e.g. `artemis.version_control.url: http://172.33.0.2` that are then reachable both from client and Jenkins container.
> 
> Now those two uses of the URLs are separated and Jenkins/Gitlab can communicate completely inside the docker network with each other via their hostnames and possibly changing IPs while the client uses their externally available URLs.
> 
> Therefore this fixes not only the problem for the local setup in docker, but also production use cases where the communication between Gitlab and Jenkins is different to the one of each of the services to the outside.
> 
> **RFC**: Should `jenkins.docker` maybe be renamed to something like `jenkins.internal-urls` in `application-jenkins.yml` with an updated comment as well to reflect that those settings may not only be useful in a local setup? Then the `JenkinsDockerUrlService` and related functions should get a new name, too.

That's exactly the issue I'm trying to fix with this PR :) Setting up the Jenkins/Gitlab on Linux worked flawlessly because for some reason I'm allowed to access the containers using their IP address. But that's not possible on Windows/Mac.

The goal I had in mind with this "fix" is to provide some configuration options in order to get the Jenkins/Gitlab environments set up quickly without having the need of configuration extra things like ngix

File: src/main/resources/config/application-jenkins.yml [hunk 5]
```diff
@@ -15,3 +15,10 @@ jenkins:
     # If activated, the user's first and last name are not transferred to jenkins.
     # Instead the user will be called "user <loginname>".
     use-pseudonyms: false
+
+    # You can override the CI and VCS urls that are used internally by Artemis.
+    internal-urls:
+        # Override the ci url used e.g in Gitlab as the websocket url.
```
Comment (b-fein) [inline src/main/resources/config/application-jenkins.yml:21]:
```suggestion
        # Override the ci url used e.g in Gitlab as the webhook url.
```

File: src/main/resources/config/application-jenkins.yml [hunk 6]
```diff
@@ -15,3 +15,10 @@ jenkins:
     # If activated, the user's first and last name are not transferred to jenkins.
     # Instead the user will be called "user <loginname>".
     use-pseudonyms: false
+
+    # You can override the CI and VCS urls that are used internally by Artemis.
```
Comment (b-fein) [inline src/main/resources/config/application-jenkins.yml:19]:
```suggestion
    # You can override the CI and VCS URLs that are used internally by Artemis.
    # This is useful when the internal communication between Jenkins and Gitlab is different
    # to how clients should access those services.
    # The client-facing URLs (e.g. for the repository clone URL, link to the build plans)
    # will still be constructed from the ones defined in application-artemis.yml.
```

**Review by b-fein — APPROVED** [review]
Looks good!
I added one more comment to explain with a bit more information when the feature is useful and one fixing a typo.

**Comment by ivanchimeno** [comment]
> Looks good!
> I added one more comment to explain with a bit more information when the feature is useful and one fixing a typo.

Thank you @b-fein for your suggestions :)

**Review by b-fein — APPROVED** [review]
Re-approving.

**Review by dfuchss — APPROVED** [review]
Code and Tests are looking good to me. Nice work.

**Pull request merged**